test/frontend/add_tear_down_test.dart

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:async/async.dart';
+import 'package:test/test.dart';
+
+import '../utils.dart';
+
+void main() {
+  test("runs after the test body", () {
+    return expectTestsPass(() {
+      var test1Run = false;
+      var tearDownRun = false;
+      test("test 1", () {
+        addTearDown(() {
+          expect(test1Run, isTrue);
+          expect(tearDownRun, isFalse);
+          tearDownRun = true;
+        });
+
+        expect(tearDownRun, isFalse);
+        test1Run = true;
+      });
+
+      test("test 2", () {
+        expect(tearDownRun, isTrue);
+      });
+    });
+  });
+
+  test("multiples run in reverse order", () {
+    return expectTestsPass(() {
+      var tearDown1Run = false;
+      var tearDown2Run = false;
+      var tearDown3Run = false;
```

Comment:
[nit] does having 3 tear downs instead of 2 add any extra safety? Would cut down on clutter if we used only 2.